pd-client/client.go

```diff
@@ -14,106 +14,320 @@
 package pd
 
 import (
+	"net"
+	"net/url"
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/juju/errors"
 	"github.com/ngaut/log"
 	"github.com/pingcap/kvproto/pkg/metapb"
-	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/pdpb2"
+	"github.com/pingcap/pd/pkg/apiutil"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
 )
 
 // Client is a PD (Placement Driver) client.
 // It should not be used after calling Close().
 type Client interface {
 	// GetClusterID gets the cluster ID from PD.
-	GetClusterID() uint64
+	GetClusterID(ctx context.Context) uint64
 	// GetTS gets a timestamp from PD.
-	GetTS() (int64, int64, error)
+	GetTS(ctx context.Context) (int64, int64, error)
 	// GetRegion gets a region and its leader Peer from PD by key.
 	// The region may expire after split. Caller is responsible for caching and
 	// taking care of region change.
 	// Also it may return nil if PD finds no Region for the key temporarily,
 	// client should retry later.
-	GetRegion(key []byte) (*metapb.Region, *metapb.Peer, error)
+	GetRegion(ctx context.Context, key []byte) (*metapb.Region, *metapb.Peer, error)
 	// GetStore gets a store from PD by store id.
 	// The store may expire later. Caller is responsible for caching and taking care
 	// of store change.
-	GetStore(storeID uint64) (*metapb.Store, error)
+	GetStore(ctx context.Context, storeID uint64) (*metapb.Store, error)
 	// Close closes the client.
 	Close()
 }
 
+type tsoRequest struct {
+	done     chan error
+	physical int64
+	logical  int64
+}
+
+const (
+	pdTimeout             = 3 * time.Second
+	maxMergeTSORequests   = 10000
+	maxInitClusterRetries = 100
+)
+
+var errTSOLength = errors.New("[pd] tso length in rpc response is incorrect")
+
 type client struct {
-	worker *rpcWorker
+	urls        []string
+	clusterID   uint64
+	tsoRequests chan *tsoRequest
+
+	mu            sync.RWMutex // TODO: use embedded struct style.
+	clients       map[string]pdpb2.PDClient
```

Comment:
When leader change is detected, we do not want to drop the old rpc client (which will cancel all pending requests). Using the map is the most convenient way I can think of.